connection.go

```diff
@@ -162,28 +165,187 @@ func (mc *mysqlConn) Prepare(query string) (driver.Stmt, error) {
+func (mc *mysqlConn) escapeBytes(buf, v []byte) []byte {
+	buf = append(buf, '\'')
+	if mc.status&statusNoBackslashEscapes == 0 {
+		buf = escapeBytesBackslash(buf, v)
+	} else {
+		buf = escapeBytesQuotes(buf, v)
+	}
+	return append(buf, '\'')
+}
+
+func (mc *mysqlConn) escapeString(buf []byte, v string) []byte {
+	buf = append(buf, '\'')
+	if mc.status&statusNoBackslashEscapes == 0 {
+		buf = escapeStringBackslash(buf, v)
+	} else {
+		buf = escapeStringQuotes(buf, v)
+	}
+	return append(buf, '\'')
+}
+
+// estimateParamLength calculates upper bound of string length from types.
+func estimateParamLength(args []driver.Value) (int, bool) {
+	l := 0
+	for _, a := range args {
+		switch v := a.(type) {
+		case int64, float64:
+			// 24 (-1.7976931348623157e+308) may be upper bound. But I'm not sure.
+			l += 25
+		case bool:
+			l += 1 // 0 or 1
+		case time.Time:
+			l += 30 // '1234-12-23 12:34:56.777777'
+		case string:
+			l += len(v)*2 + 2
+		case []byte:
+			l += len(v)*2 + 2
+		default:
+			return 0, false
+		}
+	}
+	return l, true
+}
+
+func (mc *mysqlConn) interpolateParams(query string, args []driver.Value) (string, error) {
+	estimated, ok := estimateParamLength(args)
+	if !ok {
+		return "", driver.ErrSkip
+	}
+	estimated += len(query)
+
+	buf := make([]byte, 0, estimated)
+	argPos := 0
+
+	for i := 0; i < len(query); i++ {
+		q := strings.IndexByte(query[i:], '?')
+		if q == -1 {
+			buf = append(buf, query[i:]...)
+			break
+		}
+		buf = append(buf, query[i:i+q]...)
+		i += q
+
+		arg := args[argPos]
+		argPos++
+
+		if arg == nil {
+			buf = append(buf, "NULL"...)
+			continue
+		}
+
+		switch v := arg.(type) {
+		case int64:
+			buf = strconv.AppendInt(buf, v, 10)
+		case float64:
+			buf = strconv.AppendFloat(buf, v, 'g', -1, 64)
+		case bool:
+			if v {
+				buf = append(buf, '1')
+			} else {
+				buf = append(buf, '0')
+			}
+		case time.Time:
+			if v.IsZero() {
+				buf = append(buf, "'0000-00-00'"...)
+			} else {
+				v := v.In(mc.cfg.loc)
+				year := v.Year()
+				month := v.Month()
+				day := v.Day()
+				hour := v.Hour()
+				minute := v.Minute()
+				second := v.Second()
+				micro := v.Nanosecond() / 1000
+
+				buf = append(buf, []byte{
+					byte('\''),
+					byte('0' + year/1000),
+					byte('0' + year/100%10),
+					byte('0' + year/10%10),
+					byte('0' + year%10),
+					byte('-'),
+					byte('0' + month/10),
+					byte('0' + month%10),
+					byte('-'),
+					byte('0' + day/10),
+					byte('0' + day%10),
+					byte(' '),
+					byte('0' + hour/10),
+					byte('0' + hour%10),
+					byte(':'),
+					byte('0' + minute/10),
+					byte('0' + minute%10),
+					byte(':'),
+					byte('0' + second/10),
+					byte('0' + second%10),
+				}...)
+
+				if micro != 0 {
+					buf = append(buf, []byte{
+						byte('.'),
+						byte('0' + micro/100000),
+						byte('0' + micro/10000%10),
+						byte('0' + micro/1000%10),
+						byte('0' + micro/100%10),
+						byte('0' + micro/10%10),
+						byte('0' + micro%10),
+					}...)
```

Comment:
MySQL 5.6 rounds under 1 micro second.
Should I add 500 nanosecond before formatting?